src/libraries/Common/src/System/Net/SockAddr.cs

```diff
@@ -0,0 +1,105 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+
+namespace System.Net.Sockets
+{
+    [StructLayout(LayoutKind.Explicit)]
+    internal partial struct SockAddr
+    {
+        internal const int IPv4AddressSize = 4;
+        internal const int IPv6AddressSize = 16;
+
+        [FieldOffset(0)]
+        public Interop.Sys.sockaddr_in Ipv4;
+        [FieldOffset(0)]
+        public Interop.Sys.sockaddr_in6 Ipv6;
```

Comment:
It is not just the call, it is also pinning https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Common/src/System/Net/SocketAddressPal.Unix.cs#L54-L61

for something that would be easy to do in managed code. I feel the `sockaddr` is unlikely to change for platforms and families we care about. If anything, I would keep code to get size of `sockaddr_storage`  in native so we do not need to track addition of new protocols. 

And we do it for `Quic` already anyway. I feel it would be _nice_ if we can share more code.

Reply:
Is the pinning cost worth the duplication?

If you're adding the mapping in managed code, then some existing users of `Get/SetIPv{4,6}Address` can probably also use it.

Reply:
Pinning with fixed is close to free unless there's a GC that occurs concurrently, and that's not particularly likely with a really fast call like this one. Plus if the thing being pinned isn't actually on the heap, the cost during a GC that does occur is also minimal.  I wouldn't duplicate a lot of code to avoid it.